src/client/src/offset.rs

```diff
@@ -0,0 +1,273 @@
+            return None;
+        }
+        Some(Self {
+            inner: OffsetInner::FromBeginning(offset),
+        })
+    }
+
+    /// Creates a relative offset counting backwards from the end of the log
+    ///
+    /// A relative `FromEnd` offset will begin counting from the last
+    /// "stable committed" event entry in the log. Increasing the offset will
+    /// select events in reverse chronological order from the most recent event
+    /// towards the earliest event. Therefore, a relative `FromEnd` offset may
+    /// refer to different entries depending on when a query is made.
+    ///
+    /// For example, `Offset::from_end(3)` will refer to the event with content
+    /// `33` at this point in time:
+    ///
+    /// ```text
+    ///         Partition Log: [ .., .., 22, 33, 44, 55, 66 ]
+    ///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6
+    ///        FromEnd Offset:    6,  5,  4,  3,  2,  1,  0
+    /// ```
+    ///
+    /// But when these new events are added, `Offset::from_end(3)` will refer to
+    /// the event with content `66`:
+    ///
+    /// ```text
+    ///                                        These events were added!
+    ///                                                               |
+    ///                                                      vvvvvvvvvv
+    ///         Partition Log: [ .., .., 22, 33, 44, 55, 66, 77, 88, 99 ]
+    ///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6,  7,  8,  9
+    ///        FromEnd Offset:    9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+    /// ```
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use fluvio::Offset;
+    /// assert!(Offset::from_end(10).is_some());
+    /// assert!(Offset::from_end(0).is_some());
+    /// assert!(Offset::from_end(-10).is_none());
+    /// ```
+    pub fn from_end(offset: i64) -> Option<Offset> {
+        if offset < 0 {
+            return None;
+        }
+        Some(Self {
+            inner: OffsetInner::FromEnd(offset),
+        })
+    }
+
+    /// Converts this offset into an absolute offset
+    ///
+    /// If this offset is relative from the beginning (i.e. it was created
+    /// using the [`from_beginning`] function), then using `to_absolute` will
+    /// calculate the absolute offset by finding the first still-persisted
+    /// event offset and adding the relative offset to it.
+    ///
+    /// Similarly, if this offset is relative from the end (i.e. it was created
+    /// using the [`from_end`] function), then using `to_absolute` will calculate
+    /// the absolute offset by finding the last stably-committed event and subtracting
+    /// the relative offset from it.
+    ///
+    /// Calling `to_absolute` on an offset that is already absolute just returns
+    /// that same offset.
+    ///
+    /// Note that calculating relative offsets requires connecting to Fluvio, and
+    /// therefore it is `async` and returns a `Result`.
+    pub(crate) async fn to_absolute<F, S: Into<String>>(
```

Comment:
this lead to dependencies to client.  this seems be belong to client

Reply:
I don't think it's terribly important where it goes, it's not a public function. I just thought it made sense to put a conversion function on the Offset type itself.